pkg/common/security/usergroup_ldap_resolver.go

```diff
@@ -0,0 +1,229 @@
+				zap.Error(err))
+			continue
+		}
+		secretValue := strings.TrimSpace(string(secretValueBytes))
+
+		// log the secret key and value
+		log.Log(log.Security).Debug("Loaded LDAP secret",
+			zap.String("key", secretKey))
+
+		switch secretKey {
+		case common.LdapHost:
+			ldapConf.Host = secretValue
+		case common.LdapPort:
+			ldapConf.Port, err = strconv.Atoi(secretValue)
+			if err != nil {
+				log.Log(log.Security).Warn("Failed to convert LDAP port to integer, using default port", zap.Error(err))
+				ldapConf.Port = common.DefaultLdapPort
+			}
+		case common.LdapBaseDN:
+			ldapConf.BaseDN = secretValue
+		case common.LdapFilter:
+			ldapConf.Filter = secretValue
+		case common.LdapGroupAttr:
+			ldapConf.GroupAttr = secretValue
+		case common.LdapReturnAttr:
+			ldapConf.ReturnAttr = strings.Split(secretValue, ",")
+		case common.LdapBindUser:
+			ldapConf.BindUser = secretValue
+		case common.LdapBindPassword:
+			ldapConf.BindPassword = secretValue
+		case common.LdapInsecure:
+			ldapConf.Insecure, err = strconv.ParseBool(secretValue)
+			if err != nil {
+				log.Log(log.Security).Warn("Failed to convert LDAP Insecure to boolean, using default value", zap.Error(err))
+				ldapConf.Insecure = common.DefaultLdapInsecure
+			}
+		case common.LdapSSL:
+			ldapConf.SSL, err = strconv.ParseBool(secretValue)
+			if err != nil {
+				log.Log(log.Security).Warn("Failed to convert LDAP SSL to boolean, using default value", zap.Error(err))
+				ldapConf.SSL = common.DefaultLdapSSL
+			}
+		default:
+			log.Log(log.Security).Warn("Encountered unrecognized LDAP secret key",
+				zap.String("key", secretKey))
+		}
+	}
+	log.Log(log.Security).Info("Finished loading LDAP secrets",
+		zap.Int("numberOfSecretsLoaded", secretCount))
+}
+
+func GetUserGroupCacheLdap() *UserGroupCache {
+	readSecrets()
+
+	return &UserGroupCache{
+		ugs:           map[string]*UserGroup{},
+		interval:      cleanerInterval * time.Second,
+		lookup:        LdapLookupUser,
+		lookupGroupID: LdapLookupGroupID,
+		groupIds:      LDAPLookupGroupIds,
+	}
+}
+
+// Default linux behaviour: a user is member of the primary group with the same name
+func LdapLookupUser(userName string) (*user.User, error) {
+	return &user.User{
+		Uid:      "1211",
+		Gid:      userName,
+		Username: userName,
+	}, nil
+}
+
+func LdapLookupGroupID(gid string) (*user.Group, error) {
+	group := user.Group{Gid: gid}
+	group.Name = gid
+	return &group, nil
+}
+
+func LDAPLookupGroupIds(osUser *user.User) ([]string, error) {
+	sr, err := LDAPConn_Bind(osUser.Username)
+	if err != nil {
+		return nil, err
+	}
+
+	var groups []string
+	for _, entry := range sr.Entries {
+		a := entry.GetAttributeValues("memberOf")
+		println(a)
+		for i := range a {
+			s := strings.Split(a[i], ",")
+			newgroup := strings.Split(s[0], "CN=")
+
+			groups = append(groups, newgroup[1])
+		}
+	}
+	return groups, nil
+}
+
+var LDAPConn_Bind = func(userName string) (*ldap.SearchResult, error) {
```

Comment:
I'd refactor this code for better testability.
Try to hide as much LDAP-specific calls as possible, thus making the main business logic more testable.

For example, an interface could be:
```
type LdapAccess interface {
  DialURL(..) (*Conn, error)
  Bind(...) error
  Search(...) (*SearchResult, error)
}
```

Then we can go for two implementations, a true `LdapAccessImpl` and `LdapAccessMock`.
As usual, write a helper function which creates `LdapAccessImpl`, eg.

```
func newLdapAccessImpl(LdapResolverConfig config) *LdapAccessImpl {
  ... create based on config entries ...
}
```

This way, you don't have to replace `LDAPConn_Bind` for testing, just call `newLdapAccessImpl()` in the production code and use a mock in the test code.

Reply:
Also, rename `LDAPConn_Bind()` to something better like `LdapSearch()`. Then this function can receive the `LdapAccess` implementation in the first parameter:

```
func LdapSearch(ldapAccess *LdapAccess, userName string) (*ldap.SearchResult, error) {
   ...
}
```

Obviously a different path can be chosen if that's more appropriate.